### DISCUSSION TITLE
How to setup primary keys and indexes for target schema

### DISCUSSION THREAD

**Discussion by gkarugi** [meltano/sdk#2175] [General]
Hey all,

I have been researching on how to add primary keys and indexes to the target schema. Without these indexes, the target becomes very slow if the data amount is huge.

I cannot seem to find any documentation or code sample of such a usecase on the docs, slack and github.

It seems like singer/meltano supports setting this, but currently cannot figure out how. 

see primary keys reference
[https://github.com/shaped-ai/target-clickhouse/blob/f70e4a7f92162922dae0d7936639c794a837d41e/target_clickhouse/connectors.py#L89](https://github.com/shaped-ai/target-clickhouse/blob/f70e4a7f92162922dae0d7936639c794a837d41e/target_clickhouse/connectors.py#L89)

**Comment by edgarrmondragon** [comment]
Hi @gkarugi! The tap communicates that information to the target via the catalog metadata[^1].

The docs are indeed lacking an example for that, but you can use the metadata extra[^2] to set them:

```yaml
extractors:
- name: your-tap
  metadata:
    some_stream_id:
      table-key-properties: [id]
```

We do have an issue to add examples like this to the Meltano docs: https://github.com/meltano/meltano/issues/7316

[^1]: https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#metadata
[^2]: https://docs.meltano.com/concepts/plugins/#metadata-extra

**Reply by matt-snider** [reply]
Hi @edgarrmondragon I am also investigating the same thing and `table-key-properties` seems to only be applicable for primary keys (e.g. when using a SQL target).

There does not seem to be any standard or built-in way to configure regular indexes (non-primary). I see that some targets, like [`pipelineweise/target-postgres`](https://github.com/transferwise/pipelinewise/tree/master/singer-connectors/target-postgres) will [load the `indices` from `schema_mapping`](https://github.com/transferwise/pipelinewise/blob/5b4110543a3c7677f5b95741c5f07bec384f66f3/singer-connectors/target-postgres/target_postgres/db_sync.py#L258)
and then [create the index](https://github.com/transferwise/pipelinewise/blob/5b4110543a3c7677f5b95741c5f07bec384f66f3/singer-connectors/target-postgres/target_postgres/db_sync.py#L488), but this seems to be a feature specific to those targets.

I was wondering if there is anything standard that can be done at the stream level, or any other common strategies to create indexes after the target runs.

**Reply by edgarrmondragon** [reply]
There's nothing in [the Singer spec](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#metadata) that would helps us here, though we could always extend it in a backwards-compatible way.

We could also of course add support as part of https://github.com/meltano/sdk/issues/1086, similar to Wise's targets.
